Node: Hugging Face Hub models

liyier90 opened the pull request:
Partially addresses https://github.com/aisingapore/PeekingDuck-Private/issues/31

CDR: https://www.notion.so/2b19acf4457c49b8a4ee52692ce80426?v=bcdfe42c992d427a8803921ffb954f5f

- Only "object detection" models in this PR.
  - All happen to be PyTorch models, hardcodes some PyTorch specific API due to lack of counterexamples
- Valid/supported models are chosen by the following criteria:
  1. Can called through `transformers`
     - Non-`transformers` models are poorly documented and have inconsistent interface [ex1](https://huggingface.co/SamMorgan/yolo_v4_tflite), [ex2](https://huggingface.co/johko/wideresnet28-2-mnist)
  2. Is not part of the OWL-ViT family
     - OWL-ViT models require an additional text input, using an empty string `""` gives empty detections. Inconsistent with the rest of PKD's object detection models [ref](https://huggingface.co/google/owlvit-large-patch14)
  3. Does not **explicitly** mention they have a GPL license.
     - Only using the information provided by `huggingface_hub.hf_api` package.
- After filtering, there are a number of models which cannot be used, due to:
  1. Missing `preprocessor_config.json`: the model authors failed to upload a complete set of files
  2. `MaskFormerConfig` is not supported by `AutoModel`: either wrong model tag or an in-development feature
  - No special handling of these cases, the original exception is thrown

**Deviation from other model nodes**
- No integrity check, set paths for cache directories instead
- Maps detect label to ID by itself instead of using `master_map.yml`
  - Defaults to detect all instead of "person" when list is empty as "person" is not a valid label for some datasets

**Utility features**
- CLI command `models` to list all valid models
- CLI command `detect-ids` to return the URL to `config.json` which contains the label-to-id mapping [example](https://huggingface.co/facebook/detr-resnet-50/blob/main/config.json)
- Nest these command under `model-hub`
  - Example:
    ```
    peekingduck model-hub huggingface models --task 'object detection'
    ```

ongtw commented:
On macOS, `transformers` need to be part of `requirements.txt` and not of optional requirements.
```
$ peekingduck run
[ log truncated ]
2022-09-07 14:56:47 peekingduck.declarative_loader  INFO:  Initializing model.huggingface_hub node...
2022-09-07 14:56:47 root  ERROR:  Traceback (most recent call last):
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/bin/peekingduck", line 8, in <module>
    sys.exit(cli())
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/peekingduck/commands/core.py", line 118, in run
    runner = Runner(
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/peekingduck/runner.py", line 76, in __init__
    self.pipeline = self.node_loader.get_pipeline()
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/peekingduck/declarative_loader.py", line 261, in get_pipeline
    instantiated_nodes = self._instantiate_nodes()
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/peekingduck/declarative_loader.py", line 181, in _instantiate_nodes
    instantiated_node = self._init_node(
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/peekingduck/declarative_loader.py", line 197, in _init_node
    node = importlib.import_module(path_to_node + node_name)
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
   File "<frozen importlib._bootstrap_external>", line 843, in exec_module
   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/peekingduck/pipeline/nodes/model/huggingface_hub.py", line 23, in <module>
    from peekingduck.pipeline.nodes.model.huggingface_hubv1 import huggingface_hub_model
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/peekingduck/pipeline/nodes/model/huggingface_hubv1/huggingface_hub_model.py", line 23, in <module>
    from transformers import AutoConfig, AutoFeatureExtractor, AutoModelForObjectDetection

2022-09-07 14:56:47 root  ERROR:  ModuleNotFoundError: No module named 'transformers'
```

ongtw commented:
**CLI Usage**
```
❯ peekingduck model-hub huggingface models --task
Error: Option '--task' requires an argument.
❯ peekingduck model-hub huggingface models --task --help
--help is an invalid/unsupported task.
```
Would it be possible for the user to see a list of tasks on CLI?

ongtw commented:
First `model.huggingface_hub` successful run on macOS (on source `highway_cars.mp4`)! 💪🏼
```
❯ peekingduck run
2022-09-07 15:03:15 peekingduck.declarative_loader  INFO:  Successfully loaded pipeline file.
2022-09-07 15:03:15 peekingduck.declarative_loader  INFO:  Initializing input.visual node...
2022-09-07 15:03:15 peekingduck.declarative_loader  INFO:  Config for node input.visual is updated to: 'source': https://storage.googleapis.com/peekingduck/videos/highway_cars.mp4
2022-09-07 15:03:16 peekingduck.pipeline.nodes.input.visual  INFO:  Input size: 720 by 480
2022-09-07 15:03:16 peekingduck.declarative_loader  INFO:  Initializing model.huggingface_hub node...
2022-09-07 15:03:19 peekingduck.declarative_loader  INFO:  Config for node model.huggingface_hub is updated to: 'model_type': hustvl/yolos-tiny
Downloading config.json: 100%|████████████████████████████████████| 4.04k/4.04k [00:00<00:00, 1.10MB/s]
2022-09-07 15:03:23 peekingduck.pipeline.nodes.model.huggingface_hubv1.huggingface_hub_model  WARNING:  Detecting all objects.
Downloading preprocessor_config.json: 100%|████████████████████████████| 275/275 [00:00<00:00, 111kB/s]
Downloading pytorch_model.bin: 100%|██████████████████████████████| 24.8M/24.8M [00:02<00:00, 9.51MB/s]
2022-09-07 15:03:34 peekingduck.pipeline.nodes.model.huggingface_hubv1.huggingface_hub_model  INFO:  Hugging Face model loaded with the following configs:
        Model type: hustvl/yolos-tiny
        IDs being detected: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 27, 28, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 67, 70, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90]
        IOU threshold: 0.6
        Score threshold: 0.5
        Class agnostic NMS: True
2022-09-07 15:10:46 peekingduck.declarative_loader  INFO:  Initializing draw.bbox node...
2022-09-07 15:10:46 peekingduck.declarative_loader  INFO:  Initializing output.screen node...
```

ongtw commented:
Runtime errors on some models without `preprocessor_config.json`:
```
❯ peekingduck run
2022-09-07 15:22:51 peekingduck.declarative_loader  INFO:  Successfully loaded pipeline file.
2022-09-07 15:22:51 peekingduck.declarative_loader  INFO:  Initializing input.visual node...
2022-09-07 15:22:51 peekingduck.declarative_loader  INFO:  Config for node input.visual is updated to: 'source': https://storage.googleapis.com/peekingduck/videos/highway_cars.mp4
2022-09-07 15:22:52 peekingduck.pipeline.nodes.input.visual  INFO:  Input size: 720 by 480
2022-09-07 15:22:52 peekingduck.declarative_loader  INFO:  Initializing model.huggingface_hub node...
2022-09-07 15:22:52 peekingduck.declarative_loader  INFO:  Config for node model.huggingface_hub is updated to: 'model_type': CellsInACell/faster_rcnn_count_cho_cells
Downloading config.json: 100%|███████████████████████████████████| 78.0/78.0 [00:00<00:00, 21.9kB/s]
2022-09-07 15:22:56 peekingduck.pipeline.nodes.model.huggingface_hubv1.huggingface_hub_model  WARNING:  Detecting all objects.
2022-09-07 15:22:57 root  ERROR:  Traceback (most recent call last):
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/transformers/feature_extraction_utils.py", line 405, in get_feature_extractor_dict
    resolved_feature_extractor_file = cached_path(
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/transformers/utils/hub.py", line 284, in cached_path
    output_path = get_from_cache(
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/transformers/utils/hub.py", line 502, in get_from_cache
    _raise_for_status(r)
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/transformers/utils/hub.py", line 411, in _raise_for_status
    raise EntryNotFoundError(f"404 Client Error: Entry Not Found for url: {response.url}")
 transformers.utils.hub.EntryNotFoundError: 404 Client Error: Entry Not Found for url: https://huggingface.co/CellsInACell/faster_rcnn_count_cho_cells/resolve/main/preprocessor_config.json
[log truncated]
   File "/opt/homebrew/Caskroom/miniforge/base/envs/pkd38/lib/python3.8/site-packages/transformers/feature_extraction_utils.py", line 430, in get_feature_extractor_dict
    raise EnvironmentError(

2022-09-07 15:22:57 root  ERROR:  OSError: CellsInACell/faster_rcnn_count_cho_cells does not appear to have a file named preprocessor_config.json.
```

ongtw reviewed (changes requested):
Some comments for consideration.

ongtw commented:
Was trying to get M1 MPS support to work over the weekend but failed as HF models used some ops that are currently unsupported. Enabling `PYTORCH_ENABLE_MPS_FALLBACK=1` didn't work: returned other errors, like expecting f32 tensor but got i64 instead.

liyier90 commented on peekingduck/optional_requirements.txt:
```diff
@@ -1,3 +1,5 @@
-model.jde PYTHON lap == 0.4.0
 model.fairmot PYTHON lap == 0.4.0
+model.huggingface PYTHON timm >= 0.6.7
+model.huggingface PYTHON transformers >= 4.21.2
```
Replaced

liyier90 commented on peekingduck/pipeline/nodes/model/huggingface_hubv1/api_utils.py:
```diff
@@ -0,0 +1,51 @@
+    return (
+        "transformers" in model_info.tags
+        and "owlvit" not in model_info.tags
+        and not any("gpl" in tag for tag in model_info.tags if "license:" in tag)
```
Updated to support `["detr", "yolos"]`.

ongtw reviewed (approved):
Ok to merge

liyier90 commented:
> On macOS, transformers need to be part of requirements.txt and not of optional requirements.

Fixed by renaming node name in `optional_requirements.txt`

> Would it be possible for the user to see a list of tasks on CLI?

Added a `tasks` CLI command. In addition, updated help message for `models` to show a list of supported tasks.

```
Usage: cli model-hub huggingface models [OPTIONS]

  Lists the valid/supported Hugging Face Hub models for the specified computer
  vision `task`.

Options:
  --task TEXT  The computer vision task, one of ['object detection'].
               [required]

  --help       Show this message and exit.
```

> Configuration error for zuppif/maskformer-swin-small-ade:

Indirectly addressed by explicitly setting a list of supported model base types
